### PR TITLE
Initial commit/first draft

### PR DESCRIPTION
**Pull request by cs-shivaji-kolape**
PR Title: SailPoint IdentityNow - Added new actions
PMDB ID: [21980](https://pmdb.fortinet.com/ProjectManagement/viewDocument.php?id=21980&doc_type=design)

Description:
SailPoint IdentitityNow connector supported following list of actions:
Get Accounts
Get Account Details
Get Account Activities
Get Account Activity
Get Password Info
Reset Password
Enable Account
Disable Account
Unlock Account
Grant Request
Revoke Request

Impact: None

Unit Testcase:

Install/Upgrade the Connector
Check dependency packages are install if required
Configure connector using (SERVER URL, Account Route, API Key, and SECRET Key)
Check all the playbooks are present
Check all the playbooks tags are present
Check all the playbooks should be in Info mode
Connector version and Sample playbook version should be same
No setup Available unit tested all actions using Dummy/Mock data.

### CODE REVIEW

File: sailpoint-identitynow/info.json [hunk 1]
```diff
@@ -0,0 +1,803 @@
...
+      "description": "Retrieves a list of all accounts details from the SailPoint IdentityNow system.",
+      "category": "investigation",
+      "annotation": "get_accounts",
+      "enabled": true,
+      "parameters": [
+        {
+          "title": "Filter",
+          "name": "filter",
+          "type": "text",
+          "tooltip": "Filter result based on provided filter Filtering is supported for the following fields and operators:\nid: eq, in \\n identityId: eq \\n name: eq, in \\n nativeIdentity: eq, in \\n sourceId: eq, in \\n uncorrelated: eq.Please refer this link https://developer.sailpoint.com/docs/standard_collection_parameters.html",
```
Comment (cs-sneha-bharote) [inline sailpoint-identitynow/info.json]: The tooltip should be small and descriptive.

File: sailpoint-identitynow/info.json [hunk 2]
```diff
@@ -0,0 +1,803 @@
...
+          "options": [
+            "SLIM",
+            "FULL"
+          ]
+        },
+        {
+          "title": "Limit",
+          "name": "limit",
+          "type": "integer",
+          "tooltip": "Max number of results to return.",
```
Comment (cs-sneha-bharote) [inline sailpoint-identitynow/info.json:97]: what is the default and max value of the limit?

File: sailpoint-identitynow/info.json [hunk 3]
```diff
@@ -0,0 +1,803 @@
...
+      "title": "Get Account Activities",
+      "description": "This gets a collection of account activities that satisfy the given query parameters.",
+      "category": "investigation",
+      "annotation": "get_account_activities",
+      "enabled": true,
+      "parameters": [
+        {
+          "title": "Type",
+          "name": "account_type",
+          "type": "text",
```
Comment (cs-sneha-bharote) [inline sailpoint-identitynow/info.json:187]: any example of account type?

File: sailpoint-identitynow/info.json [hunk 4]
```diff
@@ -0,0 +1,803 @@
...
+          "required": false,
+          "visible": true,
+          "editable": true
+        },
+        {
+          "title": "Requested For",
+          "name": "requested-for",
+          "type": "text",
+          "tooltip": "The identity that the activity was requested for. me indicates the current user. Mutually exclusive with regarding-identity.",
+          "description": "The identity that the activity was requested for. me indicates the current user. Mutually exclusive with regarding-identity.",
```
Comment (cs-sneha-bharote) [inline sailpoint-identitynow/info.json]: The identity that the activity was requested for. **"me"** indicates the current user. Mutually exclusive with regarding-identity.

File: sailpoint-identitynow/oauth_token_handler.py [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+        try:
+            response = requests.request(method=method, url=service_endpoint, data=payload, params=params,
+                                        headers=headers, verify=self.verify_ssl)
+            logger.debug("API Response Status Code: {0}".format(response.status_code))
+            logger.debug("API Response: {0}".format(response.text))
+            if response.ok:
+                return response.json()
+            if response.status_code in error_msgs.keys():
+                message = "{0}: {1}".format(error_msgs.get(response.status_code), response.text)
+                logger.error(message)
```
Comment (cs-sneha-bharote) [inline sailpoint-identitynow/oauth_token_handler.py]: Remove re-logging.

File: sailpoint-identitynow/operations.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+from Crypto.Cipher import PKCS1_v1_5 as Cipher_PKCS1_v1_5
+from connectors.core.connector import get_logger, ConnectorError
+from .oauth_token_handler import SailPointOAuth
+
+logger = get_logger('sailpoint-identitynow')
+
+
+class SailPointIdentityNow(object):
+    def __init__(self, config):
+        self.server_url = config.get('server_url', '').strip('/')
```
Comment (cs-sneha-bharote) [inline sailpoint-identitynow/operations.py]: Remove variables which are not used. 
if not self.server_url.startswith('https://') and not self.server_url.startswith('http://'):
            self._server_url = 'https://' + self.server_url
self.username = config.get('username')
        self.password = config.get('password')
        self.verify_ssl = config.get('verify_ssl')

File: sailpoint-identitynow/operations.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+        self.password = config.get('password')
+        self.verify_ssl = config.get('verify_ssl')
+        self.connector_info = config.pop('connector_info', '')
+        self.sp_auth = SailPointOAuth(config)
+        self.token = self.sp_auth.validate_token(config, self.connector_info)
+        self.headers = {'Content-Type': 'application/json', 'Authorization': '{}'.format(self.token)}
+
+
+def get_accounts(config, params):
+    query_string = {k: v for k, v in params.items() if v is not None and v != ''}
```
Comment (cs-sneha-bharote) [inline sailpoint-identitynow/operations.py]: You can write a function for this to remove redundant code.

File: sailpoint-identitynow/playbooks/playbooks.json [hunk 8]
```diff
@@ -0,0 +1,1105 @@
+{
+  "type": "workflow_collections",
+  "data": [
+    {
+      "@type": "WorkflowCollection",
+      "name": "Sample - SailPoint IdentityNow - 1.0.0",
+      "description": "Sample playbooks for \"SailPoint IdentityNow\" connector. If you are planning to use any of the sample playbooks in your environment, ensure that you clone those playbooks and move them to a different collection, since the sample playbook collection gets deleted during connector upgrade and delete.",
```
Comment (cs-sneha-bharote) [inline sailpoint-identitynow/playbooks/playbooks.json]: update PB description. Rename the image to "playbooks.png"

File: sailpoint-identitynow/info.json (hunk 3 above)
Comment (cs-shivaji-kolape) [reply sailpoint-identitynow/info.json:187]: This is an uncertified connector, and API Document only showing required string type.
Screenshot 2022-10-17 at 12 49 26 PM

**Review by cs-sneha-bharote — APPROVED** [review]
LGTM.

**Pull request merged**